components/web_discovery/browser/ecdh_aes.cc

```diff
@@ -0,0 +1,136 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/ecdh_aes.h"
+
+#include <array>
+
+#include "base/base64.h"
+#include "base/containers/span_writer.h"
+#include "base/logging.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+#include "crypto/random.h"
+#include "crypto/sha2.h"
+#include "third_party/boringssl/src/include/openssl/aead.h"
+#include "third_party/boringssl/src/include/openssl/ec.h"
+#include "third_party/boringssl/src/include/openssl/ec_key.h"
+#include "third_party/boringssl/src/include/openssl/ecdh.h"
+#include "third_party/boringssl/src/include/openssl/nid.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr size_t kAesKeySize = 16;
+constexpr size_t kAesTagLength = 16;
+constexpr size_t kIvSize = 12;
+constexpr size_t kKeyMaterialSize = 32;
+// P-256 field size * 2 + type byte
+constexpr size_t kComponentOctSize = 32 * 2 + 1;
+// type byte + public component + initialization vector
+constexpr size_t kEncodedPubKeyAndIv = 1 + kComponentOctSize + kIvSize;
+constexpr uint8_t kP256TypeByte = 0xea;
+
+}  // namespace
+
+AESEncryptResult::AESEncryptResult(std::vector<uint8_t> data,
+                                   std::string encoded_public_component_and_iv)
+    : data(data),
+      encoded_public_component_and_iv(encoded_public_component_and_iv) {}
+
+AESEncryptResult::~AESEncryptResult() = default;
+
+AESEncryptResult::AESEncryptResult(const AESEncryptResult&) = default;
+AESEncryptResult& AESEncryptResult::operator=(const AESEncryptResult&) =
+    default;
+
+std::optional<AESEncryptResult> DeriveAESKeyAndEncrypt(
+    base::span<uint8_t> server_pub_key,
+    base::span<uint8_t> data) {
+  bssl::UniquePtr<EC_KEY> client_private_key(
```

Comment:
Is the client private key used here ephemeral for each message sent in or is it re-used for multiple encrypted messages?

Reply:
Looks ephemeral here, but just want to confirm: https://github.com/brave/brave-core/pull/24971/files#diff-819965d56bebb6c02b9e058d9faffd8ef334c121329b1aa542c2b6aae4b8e9ceR118

Reply:
Yes, it's ephemeral

Reply:
sweet, thanks for confirming